mcs/class/corlib/System/TypeSpec.cs

```diff
@@ -81,35 +79,71 @@ internal class PointerSpec : ModifierSpec
 				type = type.MakePointerType ();
 			return type;
 		}
-#if DEBUG
 		public override string ToString () {
 			string s = "";
 			for (int i = 0; i < pointer_level; ++i)
 				s += "*";
 			return s;
 		}
-#endif
 
 	}
 
 	internal class TypeSpec
 	{
-		string name, assembly_name;
-		List<string> nested;
+		TypeIdentifier name;
+		string assembly_name;
+		List<TypeIdentifier> nested;
 		List<TypeSpec> generic_params;
 		List<ModifierSpec> modifier_spec;
 		bool is_byref;
 
-		bool HasModifiers {
+		string display_fullname; // cache
+
+		internal bool HasModifiers {
 			get { return modifier_spec != null; }
 		}
 
+		internal bool IsNested {
+			get { return nested != null && nested.Count > 0; }
+		}
+
+		internal bool IsByRef {
+			get { return is_byref; }
+		}
+
+		internal TypeName Name {
+			get { return name; }
+		}
+
+		internal IEnumerable<TypeName> Nested {
+			get {
+				if (nested != null)
+					return nested;
+				else
+					return new TypeName[] { };
+			}
+		}
+
+		internal IEnumerable<ModifierSpec> Modifiers {
+			get {
+				if (modifier_spec != null)
+					return modifier_spec;
+				else
+					return new ModifierSpec[] { };
```

Comment:
Same thing as the previous one, lets avoid the extra allocations here.

Quite a bit of corlib performance surprisingly comes from being very conservative in avoiding those
small, insignificant, allocations. In the aggregate they actually amount to be measurable.